File: tests/python/test_tokenizer.py
"""
tests/python/test_tokenizer.py

Run after building with CMake:
    cmake --build build --target cppBpe
    cp build/cppBpe*.so tests/python/   # or set PYTHONPATH=build
    pytest tests/python/ -v
"""

import sys
import os
import pytest
import tiktoken
import rustbpe
import regex as re
from collections import defaultdict

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "../../build"))

import cppBpe

def _ref_get_stats(ids):
    counts = {}
    for pair in zip(ids, ids[1:]):
        counts[pair] = counts.get(pair, 0) + 1
    return counts

def _ref_merge(ids, pair, idx):
    out, i = [], 0
    while i < len(ids):
        if i < len(ids) - 1 and ids[i] == pair[0] and ids[i + 1] == pair[1]:
            out.append(idx)
            i += 2
        else:
            out.append(ids[i])
            i += 1
    return out

def _ref_train(text, vocab_size, pattern):
    compiled = re.compile(pattern)
    chunks = re.findall(compiled, text)
    ids = [list(ch.encode("utf-8")) for ch in chunks]
    merges = {}
    for i in range(vocab_size - 256):
        stats = {}
        for chunk_ids in ids:
            _ref_get_stats.__wrapped__ = True
            for pair in zip(chunk_ids, chunk_ids[1:]):
                stats[pair] = stats.get(pair, 0) + 1
        if not stats:
            break
        pair = max(stats, key=stats.get)
        idx = 256 + i
        ids = [_ref_merge(chunk_ids, pair, idx) for chunk_ids in ids]
        merges[pair] = idx
    return merges

def _ref_encode(text, merges, pattern):
    compiled = re.compile(pattern)
    chunks = re.findall(compiled, text)
    result = []
    for chunk in chunks:
        ids = list(chunk.encode("UTF-8"))
        while len(ids) >= 2:
            stats = {pair: 0 for pair in zip(ids, ids[1:])}
            pair = min(stats, key=lambda p: merges.get(p, float("inf")))
            if pair not in merges:
                break
            ids = _ref_merge(ids, pair, merges[pair])
        result.extend(ids)
    return result


class TestConstruction:
    def test_default_tokenizer(self):
        tok = cppBpe.Tokenizer()
        assert tok.vocab_size == 256
        assert tok.get_pattern() == ""

    def test_vocab_size_after_training(self):
        tok = cppBpe.Tokenizer()
        texts = ["ab"] * 10 + ["cd"] * 5
        tok.train_from_iterator(iter(texts), vocab_size=257)
        assert tok.vocab_size == 257

class TestEncode:
    def test_encode_empty(self):
        tok = cppBpe.Tokenizer()
        assert tok.encode("") == []

    def test_encode_basic(self):
        tok = cppBpe.Tokenizer()
        texts = ["ab"] * 10
        tok.train_from_iterator(iter(texts), vocab_size=257, pattern=r"\w+")
        ids = tok.encode("ab")
        assert ids == [256]

    def test_encode_untrained_raw_bytes(self):
        """Without training, encode returns raw byte values (256 base tokens only)."""
        tok = cppBpe.Tokenizer()
        tok.train_from_iterator(iter([]), vocab_size=256)
        ids = tok.encode("hi")
        assert ids == [104, 105]

    def test_encode_roundtrip(self):
        tok = cppBpe.Tokenizer()
        texts = ["hello world "] * 20
        tok.train_from_iterator(iter(texts), vocab_size=300)
        text = "hello world"
        assert tok.decode(tok.encode(text)) == text

    def test_encode_unicode(self):
        tok = cppBpe.Tokenizer()
        tok.train_from_iterator(iter([]), vocab_size=256)
        text = "café"
        # roundtrip must work regardless of merge count
        decoded = tok.decode(tok.encode(text))
        assert decoded == text

class TestBatchEncode:
    def test_batch_encode_matches_single(self):
        tok = cppBpe.Tokenizer()
        texts = ["hi"] * 10 + ["hip"] * 5
        tok.train_from_iterator(iter(texts), vocab_size=257, pattern=r"\w+")

        single = [tok.encode(t) for t in ["hi", "hip", "hi"]]
        batch  = tok.batch_encode(["hi", "hip", "hi"])
        assert batch == single

    def test_batch_encode_empty_list(self):
        tok = cppBpe.Tokenizer()
        assert tok.batch_encode([]) == []

class TestDecode:
    def test_decode_empty(self):
        tok = cppBpe.Tokenizer()
        assert tok.decode([]) == ""

    def test_decode_raw_bytes(self):
        tok = cppBpe.Tokenizer()
        assert tok.decode([104, 105]) == "hi"

    def test_decode_invalid_token_raises(self):
        tok = cppBpe.Tokenizer()
        with pytest.raises(Exception):
            tok.decode([300])  # 300 not in vocab

class TestMergeableRanks:
    def test_base_256_tokens(self):
        tok = cppBpe.Tokenizer()
        ranks = tok.get_mergeable_ranks()
        assert len(ranks) == 256
        # first byte
        b0, r0 = ranks[0]
        assert b0 == bytes([0]) and r0 == 0
        # last byte
        b255, r255 = ranks[255]
        assert b255 == bytes([255]) and r255 == 255

    def test_merge_adds_entry(self):
        tok = cppBpe.Tokenizer()
        texts = ["ab"] * 20
        tok.train_from_iterator(iter(texts), vocab_size=257, pattern=r"\w+")
        ranks = tok.get_mergeable_ranks()
        assert len(ranks) == 257
        # last entry should be b"ab" → 256
        rank_dict = {k: v for k, v in ranks}
        assert rank_dict.get(b"ab") == 256

    def test_tiktoken_export(self):
        """Smoke-test that the output can be used by tiktoken if available."""

        tok = cppBpe.Tokenizer()
        texts = ["hello world "] * 30
        tok.train_from_iterator(iter(texts), vocab_size=300)

        enc = tiktoken.Encoding(
            name="test_enc",
            pat_str=tok.get_pattern(),
            mergeable_ranks={k: v for k, v in tok.get_mergeable_ranks()},
            special_tokens={},
        )
        ids  = enc.encode("hello world")
        text = enc.decode(ids)
        assert text == "hello world"

class TestTraining:
    def test_train_from_list(self):
        tok = cppBpe.Tokenizer()
        tok.train_from_iterator(["hello", "world", "hello"], vocab_size=260)
        assert tok.vocab_size >= 256

    def test_custom_pattern(self):
        tok = cppBpe.Tokenizer()
        texts = ["ab"] * 20
        tok.train_from_iterator(iter(texts), vocab_size=257, pattern=r"\w+")
        assert tok.get_pattern() == r"\w+"
        ids = tok.encode("ab")
        assert ids == [256]

    def test_train_chained_merges(self):
        """'aaa' × 10 with vocab 258 should produce two chained merges."""
        tok = cppBpe.Tokenizer()
        tok.train_from_iterator(["aaa"] * 10, vocab_size=258, pattern=r"\w+")
        assert tok.vocab_size == 258

        # After two merges: 'aaa' should encode to [257]
        ids = tok.encode("aaa")
        assert ids == [257]

    def test_train_most_frequent_merged_first(self):
        """The pair with the highest frequency is merged first."""
        tok = cppBpe.Tokenizer()
        # "ab" × 10, "cd" × 5 -> merge (a,b) first
        texts = ["ab"] * 10 + ["cd"] * 5
        tok.train_from_iterator(iter(texts), vocab_size=257, pattern=r"\w+")
        ranks_dict = {v: k for k, v in tok.get_mergeable_ranks()}
        assert ranks_dict[256] == b"ab"  # first merge must be 'ab'

    def test_empty_iterator_no_merges(self):
        tok = cppBpe.Tokenizer()
        tok.train_from_iterator(iter([]), vocab_size=256)
        assert tok.vocab_size == 256

    def test_buffer_size_param_ignored_gracefully(self):
        """buffer_size is accepted for API parity but does not affect results."""
        tok1 = cppBpe.Tokenizer()
        tok2 = cppBpe.Tokenizer()
        texts = ["hello world"] * 20
        tok1.train_from_iterator(iter(texts), vocab_size=260, buffer_size=4)
        tok2.train_from_iterator(iter(texts), vocab_size=260, buffer_size=65536)
        assert tok1.vocab_size == tok2.vocab_size
        
class TestCorrectnessVsReference:
    """Cross-validate cppBpe against the pure-Python reference."""
    
    UNAMBIGUOUS_TEXT = "ab" * 1000

    def test_merge_order_matches_reference(self):
        """First merge must be the same pair as the Python reference."""
        text = self.UNAMBIGUOUS_TEXT
        vocab_size = 258
        pattern = r"\w+"

        ref_merges = _ref_train(text, vocab_size, pattern)
        first_ref_bytes = bytes(next(
            p for p, idx in ref_merges.items() if idx == 256
        ))

        tok = cppBpe.Tokenizer()
        tok.train_from_iterator([text], vocab_size=vocab_size, pattern=pattern)
        ranks_dict = {v: k for k, v in tok.get_mergeable_ranks()}

        assert ranks_dict[256] == first_ref_bytes

    def test_encode_matches_reference(self):
        """Token IDs must be identical to the Python reference."""
        text = self.UNAMBIGUOUS_TEXT
        vocab_size = 266
        pattern = r"\w+"

        ref_merges = _ref_train(text, vocab_size, pattern)
        ref_ids    = _ref_encode(text[:500], ref_merges, pattern)

        tok = cppBpe.Tokenizer()
        tok.train_from_iterator([text], vocab_size=vocab_size, pattern=pattern)

        assert tok.encode(text[:500]) == ref_ids

    def test_roundtrip_matches_reference(self):
        """decode(encode(text)) must equal text, consistent with the reference."""
        text = self.UNAMBIGUOUS_TEXT[:200]
        vocab_size = 260
        pattern = r"\w+"

        ref_merges = _ref_train(text, vocab_size, pattern)
        ref_ids    = _ref_encode(text, ref_merges, pattern)

        tok = cppBpe.Tokenizer()
        tok.train_from_iterator([text], vocab_size=vocab_size, pattern=pattern)

        assert tok.encode(text) == ref_ids
        assert tok.decode(tok.encode(text)) == text
        
class TestCorrectnessVsRustbpe:
    """Cross-validate cppBpe against rustbpe."""
    
    UNAMBIGOUS_TEXT = "ab" * 1000
    REAL_TEXT = "the quick brown fox jumps over the lazy dog " * 500
    
    def test_encode_matches_rustbpe_unamigous(self):
        """Identical token ID's on strictly unamigous corpus."""
        text = self.UNAMBIGOUS_TEXT
        vocab_size = 266
        pattern = r"\w+"
        
        rust_tok = rustbpe.Tokenizer()
        rust_tok.train_from_iterator([text], vocab_size=vocab_size)
        
        cpp_tok = cppBpe.Tokenizer()
        cpp_tok.train_from_iterator([text], vocab_size=vocab_size, pattern=pattern)
        
        assert cpp_tok.encode(text[:500]) == rust_tok.encode(text[:500])
        
    def test_vocab_overlap_real_text(self):
        """100% vocab overlap on real natrual language text."""
        text = self.REAL_TEXT
        vocab_size = 512
        
        rust_tok = rustbpe.Tokenizer()
        rust_tok.train_from_iterator(text.splitlines(), vocab_size=vocab_size)
        
        cpp_tok = cppBpe.Tokenizer()
        cpp_tok.train_from_iterator(text.splitlines(), vocab_size=vocab_size)
        
        rust_vocab = {bytes(k) for k, _ in rust_tok.get_mergeable_ranks()}
        cpp_vocab = {k for k, _ in cpp_tok.get_mergeable_ranks()}
        
        overlap = len(rust_vocab & cpp_vocab) / len(rust_vocab | cpp_vocab)
        assert overlap == 1.0, f"Vocab overlap was only {overlap:.1f%}"
        
    def test_encode_matches_rustbpe_real_text(self):
        """Identical token ID's on real text at a vocab size where pairs dominate."""
        text = self.REAL_TEXT
        vocab_size = 400
        
        rust_tok = rustbpe.Tokenizer()
        rust_tok.train_from_iterator(text.splitlines(), vocab_size=vocab_size)
        
        cpp_tok = cppBpe.Tokenizer()
        cpp_tok.train_from_iterator(text.splitlines(), vocab_size=vocab_size)
        
        probe = text[:5_000]
        assert cpp_tok.encode(probe) == rust_tok.encode(probe)
        
    def test_tiktoken_export_matches_rustbpe(self):
        """tiktoken encoding from cppBpe must match rustBpe own encoding"""
        text = self.REAL_TEXT
        vocab_size = 400

        rust_tok = rustbpe.Tokenizer()
        rust_tok.train_from_iterator(text.splitlines(), vocab_size=vocab_size)
        
        cpp_tok = cppBpe.Tokenizer()
        cpp_tok.train_from_iterator(text.splitlines(), vocab_size=vocab_size)
        
        enc = tiktoken.Encoding(
            name="cpp_cross_check",
            pat_str=cpp_tok.get_pattern(),
            mergeable_ranks={k: v for k, v in cpp_tok.get_mergeable_ranks()},
            special_tokens={},
        )
        
        probe = text[:5_000]
        assert enc.encode(probe) == rust_tok.encode(probe)
        
        
class TestDeterminism:
    """Training the same data twice must procude identical results."""
    
    def test_train_twice_same_merges(self):
        tok1 = cppBpe.Tokenizer()
        tok2 = cppBpe.Tokenizer()
        texts = ["hello world"] * 50 + ["foo bar baz"] * 30
        tok1.train_from_iterator(iter(texts), vocab_size=280)
        tok2.train_from_iterator(iter(texts), vocab_size=280)
        
        ranks1 = {k: v for k, v in tok1.get_mergeable_ranks()}
        ranks2 = {k: v for k, v in tok1.get_mergeable_ranks()}
        
        assert ranks1 == ranks2
        
    def test_encode_deterministic(self):
        tok = cppBpe.Tokenizer()
        texts = ["hello world"] * 50
        tok.train_from_iterator(iter(texts), vocab_size=270)
        text = "hello world hello"
        assert tok.encode(text) == tok.encode(text)
        
class TestEdgeCases:
    """Edge cases missing from the current suite."""
    
    def test_batch_encode_with_empty_string(self):
        """A batch containing empty strings must not crash."""
        tok = cppBpe.Tokenizer()
        tok.train_from_iterator(["hello"] * 10, vocab_size=260)
        result = tok.batch_encode(["hello", "", "world", ""])
        assert result[1] == []
        assert result[3] == []
        assert result[0] == tok.encode("hello")
        assert result[2] == tok.encode("world")
        
    def test_encode_determinism(self):
        tok = cppBpe.Tokenizer()
        texts = ["hello world"] * 50
        tok.train_from_iterator(iter(texts), vocab_size=270)
        text = "hello world hello"
        assert tok.encode(text) == tok.encode(text)
        
    def test_decode_error_message(self):
        """Invalid token error should mention the bad ID."""
        tok = cppBpe.Tokenizer()
        with pytest.raises(Exception, match="300|invalid|unknown"):
            tok.decode([300])
            
    def test_retrain_resets_vocab(self):
        """Traning a second time on the same tokenizer must replace old merges."""
        tok = cppBpe.Tokenizer()
        tok.train_from_iterator(["ab"] * 20, vocab_size=257, pattern=r"\w+")
        assert tok.vocab_size == 257
        
        tok.train_from_iterator(["cd"] * 20, vocab_size=258, pattern=r"\w+")
        assert tok.vocab_size == 257
        
        ranks_dict = {k: v for k, v in tok.get_mergeable_ranks()}
        assert b"ab" not in ranks_dict
        assert b"cd" in ranks_dict

File: tests/cpp/test_tokenizer.cpp
#include "cppBpe/tokenizer.h"
#include <algorithm>
#include <cassert>
#include <iostream>
#include <functional>
#include <iostream>
#include <numeric>
#include <string>
#include <unordered_map>
#include <vector>
#include "absl/container/flat_hash_map.h"

using namespace cppBpe;

namespace
{
    int g_passed = 0;
    int g_failed = 0;
    bool g_current_failed = false;
    std::string g_current_name;

    void begin_test(const std::string& name)
    {
        g_current_name = name;
        g_current_failed = false;
    }

    void end_test()
    {
        if (g_current_failed)
        {
            std::cerr << "[FAILED]: " << g_current_name << std::endl;
            ++g_failed;
        }
        else
        {
            std::cout << "[PASSED]: " << g_current_name << std::endl;
            ++g_passed;
        }
    }

#define TEST(name) \
    do { \
        begin_test(#name); \
        name(); \
        end_test(); \
    } while (false)

#define CHECK(expr) \
    do { \
        if (!(expr)) \
        { \
            std::cerr << "  ASSERTION FAILED at " << __FILE__ << ":" << __LINE__ << ": " << #expr << std::endl; \
            g_current_failed = true; \
        } \
    } while (false)

#define CHECK_EQ(a, b) CHECK((a) == (b))
#define CHECK_NE(a, b) CHECK((a) != (b))

#define CHECK_THROW(expr) \
    do { \
        bool threw = false; \
        try { (expr); } \
        catch (...) { threw = true; } \
        if (!threw) \
        { \
            std::cerr << "  ASSERTION FAILED at " << __FILE__ << ":" << __LINE__ << ": expected exception from " << #expr << std::endl; \
            g_current_failed = true; \
        } \
    } while (false)


    Tokenizer make_tok_with_pattern(const std::string& pat, absl::flat_hash_map<Pair, TokenId, PairHash> merges = {})
    {
        Tokenizer tok(pat);
        tok.merges_ = std::move(merges);
        return tok;
    }

    void test_word_pairs()
    {
        const Word w({ 1, 2, 3, 4 });
        const std::vector<Pair> expected = { {1,2}, {2,3}, {3,4} };
        std::vector<Pair> got;
        for (std::size_t i = 0; i + 1 < w.ids.size(); ++i)
        {
            got.push_back({ w.ids[i], w.ids[i+1] });
        }
        CHECK_EQ(got, expected);
    }

    void test_word_pairs_empty()
    {
        const Word w({});
        CHECK(w.ids.size() < 2);
    }

    void test_word_pairs_single()
    {
        const Word w({ 42 });
        CHECK(w.ids.size() < 2);
    }

    void test_word_merge_pair()
    {
        // [1,2,3,1,2] merge (1,2)->99 -> [99,3,99]
        Word w({ 1, 2, 3, 1, 2 });
        w.merge_pair({ 1, 2 }, 99);
        CHECK_EQ(w.ids, (std::vector<TokenId>{ 99, 3, 99 }));
    }

    void test_word_merge_pair_adjacent()
    {
        // [1,2,1,2,1,2] -> [99,99,99]
        Word w({ 1, 2, 1, 2, 1, 2 });
        w.merge_pair({ 1, 2 }, 99);
        CHECK_EQ(w.ids, (std::vector<TokenId>{ 99, 99, 99 }));
    }

    void test_word_merge_no_match()
    {
        Word w({ 1, 2, 3 });
        auto deltas = w.merge_pair({ 4, 5 }, 99);
        CHECK_EQ(w.ids, (std::vector<TokenId>{ 1, 2, 3 }));
        const bool ok = deltas.empty() ||
                  std::ranges::all_of(deltas,
                                      [](const auto& kv) { return kv.second == 0; });
        CHECK(ok);
    }

    void test_tokenizer_new()
    {
        const Tokenizer tok;
        CHECK(tok.merges_.empty());
        CHECK(tok.get_patterns().empty());
    }

    void test_encode_with_pattern_no_merges()
    {
        // No merges, simple \w+ pattern -> raw byte values
        const auto tok = make_tok_with_pattern(R"(\w+)");
        const auto ids = tok.encode("hi");
        CHECK_EQ(ids, (std::vector<TokenId>{ 104, 105 }));
    }

    void test_encode_with_merges()
    {
        const auto tok = make_tok_with_pattern(R"(\w+)", { { {104, 105}, 256 } });

        const auto ids = tok.encode("hi");
        CHECK_EQ(ids, (std::vector<TokenId>{ 256 }));

        const auto ids2 = tok.encode("hip");
        CHECK_EQ(ids2, (std::vector<TokenId>{ 256, 112 }));
    }

    void test_get_mergeable_ranks_empty()
    {
        const Tokenizer tok;
        const auto ranks = tok.get_mergeable_ranks();
        CHECK_EQ(ranks.size(), static_cast<size_t>(256));
        // first: [0] -> 0
        CHECK_EQ(ranks[0].second, 0u);
        CHECK_EQ(ranks[0].first, (std::vector<uint8_t>{ 0 }));
        // last: [255] -> 255
        CHECK_EQ(ranks[255].second, 255u);
        CHECK_EQ(ranks[255].first, (std::vector<uint8_t>{ 255 }));
    }

    void test_get_mergeable_ranks_with_merge()
    {
        const auto tok = make_tok_with_pattern("", { { {65, 66}, 256 } });
        const auto ranks = tok.get_mergeable_ranks();
        CHECK_EQ(ranks.size(), static_cast<size_t>(257));
        CHECK_EQ(ranks[256].first, (std::vector<uint8_t>{ 65, 66 }));
        CHECK_EQ(ranks[256].second, 256u);
    }

    void test_count_pairs_parallel()
    {
        const std::vector<Word> words  = { Word({1,2,3}), Word({1,2,4}) };
        const std::vector<int32_t> counts = { 1, 2 };

        PairCounts    pc;
        WhereToUpdate wu;
        count_pairs_parallel(words, counts, pc, wu);

        // (1,2): 1*1 + 1*2 = 3
        CHECK_EQ(pc.at({1,2}), 3);
        // (2,3): 1
        CHECK_EQ(pc.at({2,3}), 1);
        // (2,4): 2
        CHECK_EQ(pc.at({2,4}), 2);

        const auto& v = wu.at({1,2});
        CHECK(std::find(v.begin(), v.end(), 0) != v.end());
        CHECK(std::find(v.begin(), v.end(), 1) != v.end());
    }

    void test_train_core_incremental()
    {
        Tokenizer tok(R"(\w+)");

        std::vector<std::string> texts(10, "ab");
        texts.insert(texts.end(), 5, "cd");
        tok.train(texts, 257);

        CHECK_EQ(tok.merges_.size(), static_cast<size_t>(1));
        CHECK(tok.merges_.count({97,98}));
        CHECK_EQ(tok.merges_.at({97,98}), 256u);
    }

    void test_default_vocab_size()
    {
        Tokenizer tok;
        CHECK_EQ(tok.vocab_size(), 256u);
        tok.merges_[{97,98}] = 256;
        CHECK_EQ(tok.vocab_size(), 257u);
    }

    void test_word_merge_overlapping_odd()
    {
        // "aaa" = [97,97,97] -> [256, 97]
        Word w({ 97, 97, 97 });
        w.merge_pair({ 97, 97 }, 256);
        CHECK_EQ(w.ids, (std::vector<TokenId>{ 256, 97 }));
    }

    void test_word_merge_overlapping_even()
    {
        // "aaaa" = [97,97,97,97] -> [256,256]
        Word w({ 97, 97, 97, 97 });
        w.merge_pair({ 97, 97 }, 256);
        CHECK_EQ(w.ids, (std::vector<TokenId>{ 256, 256 }));
    }

    void test_word_merge_multiple_occurrences()
    {
        Word w({ 1, 2, 99, 1, 2 });
        auto deltas = w.merge_pair({ 1, 2 }, 256);
        CHECK_EQ(w.ids, (std::vector<TokenId>{ 256, 99, 256 }));

        // (1,2) should have aggregate delta of -2
        int32_t ab_delta = 0;
        for (auto& [p, d] : deltas)
        {
            if (p == Pair{1, 2})
            {
                ab_delta += d;
            }
        }
        CHECK_EQ(ab_delta, -2);
    }

    void test_encode_chained_merges()
    {
        //std::unordered_map<Pair, TokenId, PairHash> m;
        absl::flat_hash_map<Pair, TokenId, PairHash> m;
        m[{97,97}]  = 256; // 'aa' -> 256 (learned first)
        m[{256,97}] = 257; // 'aa'+'a' -> 257 (learned second)
        const auto tok = make_tok_with_pattern(R"(\w+)", m);

        // "aaa" -> [257]
        CHECK_EQ(tok.encode("aaa"),  (std::vector<TokenId>{ 257 }));
        // "aaaa" -> [256, 256]
        CHECK_EQ(tok.encode("aaaa"), (std::vector<TokenId>{ 256, 256 }));
        // "aaaaa" -> [256, 257]
        CHECK_EQ(tok.encode("aaaaa"), (std::vector<TokenId>{ 256, 257 }));
    }

    void test_encode_decode_roundtrip_simple()
    {
        const auto tok = make_tok_with_pattern(R"(\w+|\s+)", { { {104,105}, 256 } });
        const auto ids = tok.encode("hi");
        const auto decoded = tok.decode(ids);
        CHECK_EQ(decoded, "hi");
    }

    void test_encode_decode_roundtrip_with_spaces()
    {
        //std::unordered_map<Pair, TokenId, PairHash> m;
        absl::flat_hash_map<Pair, TokenId, PairHash> m;
        m[{104,101}] = 256; // 'he'
        m[{108,108}] = 257; // 'll'
        m[{256,257}] = 258; // 'hell'
        const auto tok = make_tok_with_pattern(R"(\w+|\s+)", m);

        const std::string text = "hello world";
        const auto ids  = tok.encode(text);
        const auto decoded = tok.decode(ids);
        CHECK_EQ(decoded, text);
    }

    void test_decode_byte_level()
    {
        const Tokenizer tok;
        // [104, 105] = "hi"
        const auto decoded = tok.decode({ 104, 105 });
        CHECK_EQ(decoded, "hi");
    }

    void test_decode_invalid_token()
    {
        const Tokenizer tok;
        CHECK_THROW(tok.decode({ 300 }));
    }

    void test_train_creates_chained_merges()
    {
        // "aaa" ×10 -> first merge (97,97)->256, second (256,97)->257
        Tokenizer tok(R"(\w+)");
        const std::vector<std::string> texts(10, "aaa");
        tok.train(texts, 258);

        CHECK_EQ(tok.merges_.size(), static_cast<size_t>(2));
        CHECK(tok.merges_.count({97,97}));
        CHECK_EQ(tok.merges_.at({97,97}), 256u);
        CHECK(tok.merges_.count({256,97}));
        CHECK_EQ(tok.merges_.at({256,97}), 257u);
    }

    void test_get_mergeable_ranks_chained()
    {
        //std::unordered_map<Pair, TokenId, PairHash> m;
        absl::flat_hash_map<Pair, TokenId, PairHash> m;
        m[{65,66}]  = 256; // "AB"
        m[{256,67}] = 257; // "ABC"
        const auto tok = make_tok_with_pattern("", m);

        const auto ranks = tok.get_mergeable_ranks();
        CHECK_EQ(ranks.size(), static_cast<size_t>(258));
        CHECK_EQ(ranks[256].first,  (std::vector<uint8_t>{ 65, 66 }));
        CHECK_EQ(ranks[256].second, 256u);
        CHECK_EQ(ranks[257].first,  (std::vector<uint8_t>{ 65, 66, 67 }));
        CHECK_EQ(ranks[257].second, 257u);
    }

    void test_encode_empty_string()
    {
        const auto tok = make_tok_with_pattern(R"(\w+)");
        CHECK(tok.encode("").empty());
    }

    void test_encode_no_matches()
    {
        const auto tok = make_tok_with_pattern(R"(\w+)");
        CHECK(tok.encode("   ").empty());
    }

    void test_decode_empty()
    {
        const Tokenizer tok;
        CHECK_EQ(tok.decode({}), "");
    }

    void test_word_merge_deltas_correctness()
    {
        // Word: [1,2,3,1,2] merge (1,2)->99
        // Expected aggregate deltas: (1,2)=-2, (2,3)=-1, (3,1)=-1, (99,3)=+1, (3,99)=+1
        Word w({ 1, 2, 3, 1, 2 });
        auto deltas = w.merge_pair({ 1, 2 }, 99);

        std::unordered_map<Pair, int32_t, PairHash> dm;
        for (auto& [p, d] : deltas)
        {
            dm[p] += d;
        }

        CHECK_EQ(dm.at({1,2}), -2);
        CHECK_EQ(dm.at({2,3}), -1);
        CHECK_EQ(dm.at({3,1}), -1);
        CHECK_EQ(dm.at({99,3}), +1);
        CHECK_EQ(dm.at({3,99}), +1);
    }

    void test_count_pairs_parallel_empty()
    {
        PairCounts pc;
        WhereToUpdate wu;
        count_pairs_parallel({}, {}, pc, wu);
        CHECK(pc.empty());
        CHECK(wu.empty());
    }

    void test_count_pairs_parallel_zero_count()
    {
        const std::vector<Word> words = { Word({1,2,3}) };
        const std::vector<int32_t> counts = { 0 };
        PairCounts pc;
        WhereToUpdate wu;
        count_pairs_parallel(words, counts, pc, wu);
        CHECK(pc.empty());
    }

    void test_batch_encode_parallel()
    {
        const auto tok = make_tok_with_pattern(R"(\w+|\s+)", { { {104,105}, 256 } });
        const auto results = tok.batch_encode({ "hi", "hip", "hi" });
        CHECK_EQ(results.size(), static_cast<size_t>(3));
        CHECK_EQ(results[0], (std::vector<TokenId>{ 256 }));
        CHECK_EQ(results[1], (std::vector<TokenId>{ 256, 112 }));
        CHECK_EQ(results[2], (std::vector<TokenId>{ 256 }));
    }

    void test_unicode_pattern_gpt4()
    {
        // The GPT-4 pattern must match Unicode letters correctly.
        // "café" should split into chunks that include 'é' (U+00E9, 2 bytes UTF-8).
        const Tokenizer tok{std::string(GPT4_PATTERN)};
        const auto ids = tok.encode("café");
        // Roundtrip must work
        const auto decoded = tok.decode(ids);
        CHECK_EQ(decoded, "café");
    }
}

int main()
{
    try
    {
        TEST(test_word_pairs);
        TEST(test_word_pairs_empty);
        TEST(test_word_pairs_single);
        TEST(test_word_merge_pair);
        TEST(test_word_merge_pair_adjacent);
        TEST(test_word_merge_no_match);
        TEST(test_tokenizer_new);
        TEST(test_encode_with_pattern_no_merges);
        TEST(test_encode_with_merges);
        TEST(test_get_mergeable_ranks_empty);
        TEST(test_get_mergeable_ranks_with_merge);
        TEST(test_count_pairs_parallel);
        TEST(test_train_core_incremental);
        TEST(test_default_vocab_size);
        TEST(test_word_merge_overlapping_odd);
        TEST(test_word_merge_overlapping_even);
        TEST(test_word_merge_multiple_occurrences);
        TEST(test_encode_chained_merges);
        TEST(test_encode_decode_roundtrip_simple);
        TEST(test_encode_decode_roundtrip_with_spaces);
        TEST(test_decode_byte_level);
        TEST(test_decode_invalid_token);
        TEST(test_train_creates_chained_merges);
        TEST(test_get_mergeable_ranks_chained);
        TEST(test_encode_empty_string);
        TEST(test_encode_no_matches);
        TEST(test_decode_empty);
        TEST(test_word_merge_deltas_correctness);
        TEST(test_count_pairs_parallel_empty);
        TEST(test_count_pairs_parallel_zero_count);
        TEST(test_batch_encode_parallel);
        TEST(test_unicode_pattern_gpt4);
    }
    catch (const std::exception& ex)
    {
        std::cerr << "UNEXPECTED EXCEPTION: " << ex.what() << std::endl;
        return 1;
    }
    catch (...)
    {
        std::cerr << "UNEXPECTED UNKNOWN EXCEPTION" << std::endl;
        return 1;
    }

    std::cout << "Tests passed: " << g_passed << ", failed: " << g_failed << std::endl;
    return g_failed == 0 ? 0 : 1;
}